### PR TITLE
estimate fee for bulk transactions

### PR DESCRIPTION
**Pull request by lana-shanghai**
# Pull Request type

- Feature
Allow to pass a vector of transactions for fee estimation. Needed for instance for compatibility with the Argent wallet. 

## What is the current behavior?
Currently we accept only 1 transaction to estimate the fee and the state rolls back to initial state after each estimation. 
Resolves: [issue](https://github.com/keep-starknet-strange/madara/issues/1202)

## What is the new behavior?

We pass a vector of transaction to the `estimate_fee` function

### CODE REVIEW

**Review by EvolveArt — APPROVED** [review]

File: crates/client/rpc/src/lib.rs [hunk 1]
```diff
@@ -619,27 +619,33 @@ where
...
-                    error!("Request parameters error: {e}");
-                    StarknetRpcApiError::InternalServerError
-                })?
-                .map_err(|e| {
-                    error!("Failed to call function: {:#?}", e);
-                    StarknetRpcApiError::ContractError
-                })?;
-
-            estimates.push(FeeEstimate { gas_price: 0, gas_consumed: gas_usage, overall_fee: actual_fee });
+            transactions.push(tx);
```
Comment (tdelabro) [inline crates/client/rpc/src/lib.rs:628]: You can use `into_iter` and `map`

File: crates/pallets/starknet/src/lib.rs [hunk 2]
```diff
@@ -1092,64 +1090,82 @@ impl<T: Config> Pallet<T> {
...
             disable_nonce_validation: bool,
-        ) -> TransactionExecutionResult<TransactionExecutionInfo> {
-            // TODO: initialization can probably be skiped by using mem::MaybeUninit
-            let mut execution_result = Ok(Default::default());
+            chain_id: Felt252Wrapper,
+        ) -> Vec<TransactionExecutionResult<TransactionExecutionInfo>> {
+            let mut execution_results = vec![];
             let _: Result<_, DispatchError> = storage::transactional::with_transaction(|| {
-                execution_result = tx.execute(state, block_context, true, disable_nonce_validation);
+                for tx in txs {
```
Comment (tdelabro) [inline crates/pallets/starknet/src/lib.rs:1104]: You should use an iterator and map, and stop iterating at the first error encountered.
`collect::<Result<Value, E>>` allows you to do this.

Here you keep executing every transaction, even after a previous one failed. This is a waste of computation

Comment (tdelabro) [reply crates/pallets/starknet/src/lib.rs:1104]: And you just do `exectuction_results = txs.into_iter().map(|tx| {...})` to bring the result outside the closure

File: crates/pallets/starknet/src/lib.rs [hunk 3]
```diff
@@ -1092,64 +1090,82 @@ impl<T: Config> Pallet<T> {
...
+        for res in execution_results {
+            match res {
+                Ok(tx_exec_info) => {
+                    log!(info, "Successfully estimated fee: {:?}", tx_exec_info);
+                    if let Some(l1_gas_usage) = tx_exec_info.actual_resources.0.get("l1_gas_usage") {
+                        results.push((tx_exec_info.actual_fee.0 as u64, *l1_gas_usage as u64));
+                    } else {
+                        return Err(Error::<T>::TransactionExecutionFailed.into());
+                    }
+                }
```
Comment (tdelabro) [inline crates/pallets/starknet/src/lib.rs:1161]: This is not needed anymore if you stop executing the tx after the first that failed

File: crates/pallets/starknet/src/runtime_api.rs [hunk 4]
```diff
@@ -46,7 +46,7 @@ sp_api::decl_runtime_apis! {
         /// Returns the chain id.
         fn chain_id() -> Felt252Wrapper;
         /// Returns fee estimate
```
Comment (tdelabro) [inline crates/pallets/starknet/src/runtime_api.rs:48]: Add in the doc, which `u64` is the `overall_fee` and which is the `gas_consumed`

Comment (tdelabro) [reply crates/pallets/starknet/src/runtime_api.rs:48]: But even better would be creating a dedicated struct so the API users are less prone to commit mistakes

**Pull request merged**